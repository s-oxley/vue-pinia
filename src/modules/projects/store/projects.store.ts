import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import type { Project, Task } from '../interfaces/project.interface';
import { useLocalStorage } from '@vueuse/core';

export const useProjectStore = defineStore('projects', () => {
  // state
  const projects = ref(useLocalStorage<Project[]>('projects', []));

  // getters
  // const getProjects = computed(() => [...projects.value]);

  const getProjects = computed(() => projects.value);
  //lo que se agrega al array desde el getProjects queda reactivo.

  // actions
  function addProject(name: string, version: number, description: string, tasks: Task[]): void {
    const newProject: Project = {
      name,
      version,
      description,
      tasks,
    };
    projects.value.push(newProject);
  }

  return { getProjects, addProject };
});
