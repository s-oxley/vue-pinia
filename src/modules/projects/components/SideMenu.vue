<template>
  <aside class="bg-base-200 min-h-screen w-72">
    <h2 class="txt-lg font-bold mx-4">Proyectos</h2>
    <p v-if="projectStore.projectsAreEmpty" class="text-sm text-gray-500 mx-4">No hay proyectos</p>
    <!-- Menu -->
    <ul v-else class="menu">
      <li v-for="(project, index) in projectStore.getProjects" :key="index">
        <template v-if="project.tasks.length === 0">
          <RouterLink :to="`/project/${project.id}`">{{ project.name }}</RouterLink>
        </template>
        <template v-else>
          <details close>
            <summary>
              <RouterLink :to="`/project/${project.id}`">{{ project.name }}</RouterLink>
            </summary>
            <ul>
              <li v-for="(task, index) in project.tasks" :key="index">
                <RouterLink :to="`/project/${project.id}`">{{ task.name }}</RouterLink>
              </li>
            </ul>
          </details>
        </template>
      </li>
    </ul>
  </aside>
</template>

<script setup lang="ts">
import { useProjectStore } from '@/modules/projects/store/projects.store';
import { RouterLink } from 'vue-router';

const projectStore = useProjectStore();
</script>
