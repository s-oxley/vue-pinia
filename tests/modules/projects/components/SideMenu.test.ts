import SideMenu from '@/modules/projects/components/SideMenu.vue';
import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { useProjectStore } from '@/modules/projects/store/projects.store';
import { projectsFake } from '../../../mockData/projectsFake';
import { nextTick } from 'vue';

// libreria de desarrollo para testing de Pinia: https://pinia.vuejs.org/cookbook/testing.html

describe('<SideMenu />', () => {
  const wrapper = mount(SideMenu, {
    global: {
      plugins: [createTestingPinia()],
      stubs: ['router-link'],
    },
  });
  const store = useProjectStore();

  beforeEach(() => {
    store.$patch(() => {
      store.projects = [];
    });
  });

  test('should render with no projects', () => {
    // console.log(wrapper.html());
    expect(wrapper.html()).toContain('No hay proyectos');
  });

  test('should render with  projects', async () => {
    // modifica el estado del store con la funcion $patch
    store.$patch({
      projects: projectsFake,
    });

    // espera el proximo renderizado de Vue y entonces carga el cambio en el store de pinia con los datos
    // del projectsFake
    await nextTick();

    expect(wrapper.html()).not.toContain('No hay proyectos');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
