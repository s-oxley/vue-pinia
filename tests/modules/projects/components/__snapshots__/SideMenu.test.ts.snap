// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SideMenu /> > should render with  projects 1`] = `
"<aside class="bg-base-200 min-h-screen w-72">
  <h2 class="txt-lg font-bold mx-4">Proyectos</h2><!-- Menu -->
  <ul class="menu">
    <li>
      <details close="">
        <summary>
          <router-link-stub to="/project/2fd9f892-515e-466a-8f48-8f8c336dc174" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/2fd9f892-515e-466a-8f48-8f8c336dc174" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/2fd9f892-515e-466a-8f48-8f8c336dc174" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/2fd9f892-515e-466a-8f48-8f8c336dc174" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
    <li>
      <router-link-stub to="/project/afa29287-9814-479d-9f9b-b2cecb33d699" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
    </li>
    <li>
      <router-link-stub to="/project/5dbb23b7-1baa-4ff4-bee7-dcd3e68d9baf" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
    </li>
    <li>
      <router-link-stub to="/project/30e613fc-d527-40b3-a29e-5731cc6890f6" replace="false" custom="false" ariacurrentvalue="page" viewtransition="false"></router-link-stub>
    </li>
  </ul>
</aside>"
`;
